File: src/main/java/dev/cernavskis/moose/lexer/Token.java
package dev.cernavskis.moose.lexer;

import dev.cernavskis.moose.util.DebugInfo;

public final class Token {
  private final TokenType type;
  private final String value;
  private final int line;
  private final int column;
  private final String file;


  public Token(TokenType type, String value, int line, int column, String file) {
    this.type = type;
    this.value = value;
    this.line = line;
    this.column = column;
    this.file = file;
  }

  @Override
  public String toString() {
    return String.format("%s" + " ".repeat(TokenType.LONGEST_TOKEN_NAME - type.name().length()) + " %s", type.name(), value);
  }

  public TokenType type() {
    return type;
  }

  public String value() {
    return value;
  }

  public DebugInfo debugInfo() {
    return new DebugInfo(line, column, file, this);
  }
}

File: src/lexer/token.rs
use crate::util::DebugInfo;
use super::TokenType;

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub value: String,
    pub line: usize,
    pub column: usize,
    pub file: String,
}

impl Token {
    pub fn new(token_type: TokenType, value: String, line: usize, column: usize, file: String) -> Self {
        Token {
            token_type,
            value,
            line,
            column,
            file,
        }
    }

    pub fn debug_info(&self) -> DebugInfo {
        DebugInfo::new(self.line, self.column, self.file.clone(), Some(self.clone()))
    }
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let name = self.token_type.name();
        let spaces = " ".repeat(TokenType::LONGEST_TOKEN_NAME - name.len());
        write!(f, "{}{} {}", name, spaces, self.value)
    }
}
